### PR TITLE
lp1445885: Fix handling of files with wrong suffix

### PR DESCRIPTION
**Pull request by uklotzde**
The legacy code still contained lots of inconsistencies between *file type* and *file suffix* caused by the weak string typing. I have decided to adopt "suffix" instead of "extension" because `QMimeDatabase`/`QFileInfo` also use this terminology.

A new test has been added for the related MIME type detection. This revealed a bug when trying to import metadata from a missing file which resulted in another new regression test.

Please note, that the type of .mp3 files with DASH/AAC content is not updated and remains as "mp3" because `QMimeDatabase` does not recognize the actual content type! See the corresponding warning log messages.

No backport to 2.3, not worth the hassle.

#7970

### CODE REVIEW

File: src/sources/soundsourceproxy.cpp
Comment (daschuer) [inline src/sources/soundsourceproxy.cpp:295]: Why do you  think we will have two extensions per type? Remove? Else you may do two loops one counting and one appending.

File: src/sources/soundsourceproxy.cpp
Comment (daschuer) [inline src/sources/soundsourceproxy.cpp:300]: do we want to handle files without a suffix? My guess is that this is a common case on windows where the file extension is hidden.

File: src/sources/soundsourceproxy.cpp
Comment (daschuer) [inline src/sources/soundsourceproxy.cpp:659]: Nice workaround for the pair.
 
Can we rename to sourceSynchronizedAt to something that reveals that it is a QDateTime?

File: src/test/coverartutils_test.cpp
Comment (daschuer) [inline src/test/coverartutils_test.cpp:91]: mod and ogg are critical and deserves a test IMHO

File: src/test/directorydaotest.cpp
Comment (daschuer) [inline src/test/directorydaotest.cpp:36]: should this be fileType?

File: src/test/directorydaotest.cpp
Comment (daschuer) [inline src/test/directorydaotest.cpp:41]: What is special of the first?

File: src/track/serato/tags.cpp
Comment (daschuer) [inline src/track/serato/tags.cpp:24]: I think SoundSource::getTypeFromFile() does not fit to the rest of the code now. We can't be sure that the underlying magic number libraries returns the file preferedSuffix where we have a provider for. We don't have it under our control. I can change at any time when it is updated. I think we need to return all file suffixes and probe all to be sure.

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you very much. This is a big improvement! However I have a few comments.

File: src/sources/metadatasourcetaglib.cpp
Comment (uklotzde) [reply src/sources/metadatasourcetaglib.cpp:89]: ???

File: src/sources/soundsourceproxy.cpp
Comment (uklotzde) [reply src/sources/soundsourceproxy.cpp:659]: I disagree. The name does not need to repeat the type in a strongly types language.

File: src/test/coverartutils_test.cpp
Comment (uklotzde) [reply src/test/coverartutils_test.cpp:91]: Unrelated. I didn't change this code and won't extend the scope of this PR.

File: src/sources/soundsourceproxy.cpp
Comment (uklotzde) [reply src/sources/soundsourceproxy.cpp:302]: We probably never handled files without a suffix correctly and never tested it. I don't want to extend the scope of this PR.

File: src/track/serato/tags.cpp
Comment (uklotzde) [reply src/track/serato/tags.cpp:24]: I have no idea what you ask from me.

File: src/sources/metadatasourcetaglib.cpp
Comment (uklotzde) [reply src/sources/metadatasourcetaglib.cpp:89]: Files are one kind of source, currently the only one supported. What's the issue with this private utility function that returns a tuple? I just added an assertion to catch programming errors.

Comment (daschuer) [reply src/sources/metadatasourcetaglib.cpp:89]: The issue is that the code is hard to understand. I was trying to make suggestions to improve it. I actually don't mind how. How about this:
```suggestion
    const QDateTime fileSynchronizedAt =
```

File: src/sources/soundsourceproxy.cpp
Comment (daschuer) [reply src/sources/soundsourceproxy.cpp:302]: OK

File: src/sources/soundsourceproxy.cpp
Comment (daschuer) [reply src/sources/soundsourceproxy.cpp:662]: The issue here is that the type is out of sight and the variable is IMHO badly named. The reader has no hint what happens here. He has to reference other files to get an idea.

How about adding a brief comment instead?

File: src/test/coverartutils_test.cpp
Comment (daschuer) [reply src/test/coverartutils_test.cpp:91]: You change the code around file extension handling here and @Holzhaus in https://github.com/mixxxdj/mixxx/pull/4602 At one point it is IMHO mandatory that we have a test for these cases I consider as critical. 

Isn't it just a copy and paste task here?

File: src/sources/metadatasourcetaglib.cpp
Comment (uklotzde) [reply src/sources/metadatasourcetaglib.cpp:89]: I disagree, because the naming the current name reflects the name in the database schema.

Comment (uklotzde) [reply src/sources/metadatasourcetaglib.cpp:89]: I am not going to rename existing things if not necessary, out of scope.

File: src/sources/soundsourceproxy.cpp
Comment (uklotzde) [reply src/sources/soundsourceproxy.cpp:662]: I am not going to respond to "type is out of sight" comments. This change is an improvement to the code before which also didn't repeat the types.

Comment (uklotzde) [reply src/sources/soundsourceproxy.cpp:662]: If I am not allowed to write reasonably modern C++ code I will stop contributing.

File: src/test/coverartutils_test.cpp
Comment (uklotzde) [reply src/test/coverartutils_test.cpp:91]: Just checked: ogg is already tested, mod does not support cover art. What are you asking for here?

File: src/sources/soundsourceproxy.cpp
Comment (uklotzde) [reply src/sources/soundsourceproxy.cpp:662]: https://en.cppreference.com/w/cpp/language/structured_binding

File: src/sources/metadatasourcetaglib.cpp
Comment (daschuer) [reply src/sources/metadatasourcetaglib.cpp:89]: > I disagree, because the current name reflects the name in the database schema

Ah, I see. So the name is OK.

File: src/sources/soundsourceproxy.cpp
Comment (daschuer) [reply src/sources/soundsourceproxy.cpp:662]: As I have written initially I like the solution around the anonymous pair issue. This was serious. I have now also understand why the variable is named as it is. But the original issue of the hard to understand code still applies. I will propose an improved comment, which is probably the best solution.

File: src/sources/soundsourceproxy.cpp
Comment (daschuer) [inline src/sources/soundsourceproxy.cpp:661]:
```suggestion
    // Parse the tags stored in the audio file and the date and time when the
    // file has been last modified to detect future changes of the tags
```

File: src/test/coverartutils_test.cpp
Comment (daschuer) [reply src/test/coverartutils_test.cpp:91]: Ah Ok I did not know this. But it looks like we have also nowhere else a test with ```SoundSourceProxy::isFileSuffixSupported(QStringLiteral("mod"))``` 
Or "mod" in general l. 

Maybe we can put it here anyway with no cover expected? 

But the original issue we need to test if the magicNumber -> mimeType -> preeredSuffix -> fileType -> suffix round trip still works.  
Not sure if this is already covered by a test or if we need to check it explicit. 

I will explain the details in the other thread below.

File: src/track/serato/tags.cpp
Comment (daschuer) [reply src/track/serato/tags.cpp:24]: SoundSource::getTypeFromFile does the flowing:

* Look up the magic number in the file content 
* pick the mime type matching to this magic number 
* Return 1 of many suffixes which is considered as preferred.
* Return this preferred suffix as file type, even if we don't have such file type, and even if the original suffix was already matching a file type. 

This works only by a chance for all our sound sources and may break at any time. I have most concerned with ogg an mod.

A workaround would be to return all suffixed for this magic number and look, if one of them is a fileType. 

To secure this against regression that might happen due to an update of the magic numbers database, we should test these cases as a unit-test. Or verify that this is already covered by a unit-test.  

A much better workaround would be to replace the  QString fileTypeg by a MimeType value.

File: src/sources/soundsourceoggvorbis.cpp
Comment (uklotzde) [reply src/sources/soundsourceoggvorbis.cpp:35]: https://github.com/mixxxdj/mixxx/pull/4608/commits/e40fab905e5d3a84e3f9c63bfbda924f90e0545f

File: src/test/coverartutils_test.cpp
Comment (uklotzde) [reply src/test/coverartutils_test.cpp:91]: https://github.com/mixxxdj/mixxx/pull/4608/commits/e40fab905e5d3a84e3f9c63bfbda924f90e0545f

File: src/sources/soundsourcemodplug.cpp
Comment (uklotzde) [reply src/sources/soundsourcemodplug.cpp:22]: https://github.com/mixxxdj/mixxx/pull/4608/commits/e40fab905e5d3a84e3f9c63bfbda924f90e0545f

File: src/track/serato/tags.cpp
Comment (uklotzde) [reply src/track/serato/tags.cpp:24]: I am not going to replace the legacy string types, out of scope of this PR.

Comment (uklotzde) [reply src/track/serato/tags.cpp:24]: https://github.com/mixxxdj/mixxx/pull/4608/commits/e40fab905e5d3a84e3f9c63bfbda924f90e0545f

Comment (uklotzde) [reply src/track/serato/tags.cpp:24]: Please reopen if you need more tests or otherwise provide those additional tests that you want to see.

File: src/sources/soundsourceproxy.cpp
Comment (uklotzde) [reply src/sources/soundsourceproxy.cpp]: https://github.com/mixxxdj/mixxx/pull/4608/commits/adf9fafdedd329c6d66a33af66575a901b868717

**Comment by uklotzde** [comment]
The test added in https://github.com/mixxxdj/mixxx/pull/4608/commits/e40fab905e5d3a84e3f9c63bfbda924f90e0545f succeeds locally on Linux built with Qt 5.15.

File: src/sources/soundsourceproxy.cpp
Comment (Holzhaus) [reply src/sources/soundsourceproxy.cpp:663]: > The issue here is that the type is out of sight

I think using `auto` is okay. It's just a local variable. Your IDE probably shows the type when hovering. And even if it doesn't, the compiler will tell you what type it is when you try to use it.

Comment (daschuer) [reply src/sources/soundsourceproxy.cpp:663]: The issue is already fixed by the additional comment. Thank you.

File: src/track/serato/tags.cpp
Comment (daschuer) [reply src/track/serato/tags.cpp:24]: The issue pending issue is the following: 
* if a file with a magic number is for instance a "Ogg Vorbis Codec compressed Multimedia file"
* Possible extensions are "OGA, OGG, OGV, OGX"
* Qt returns one of them, based on an underlying system library not under our control. 
* By luck it will return OGG, But it might one of the others. 
* Result OGG files are not playbable anymore. This can happen due to an OS update. 

I consider this as a rock solid solution: 

Fetch all extensions from QT: Which are "OGA, OGG, OGV, OGX" and check them one by one for a sound-source. 
So the second Extension will a match and the problem is solved. 

We also need to makes sure that the FileType is considered as OGA and not OGG. 

An alternative solution would be to maintain our own list of preferred extensions for each MimeType.

Comment (uklotzde) [reply src/track/serato/tags.cpp:24]: Does this PR introduce a regression to the current state? If yes then please clarify how. If not then please merge it as.

I am not interested in extending the scope of this PR. It was only intended to fix the inconsistencies, not to improve the questionable *file type* concept that should better be replaced entirely by MIME types.

Comment (uklotzde) [reply src/track/serato/tags.cpp:24]: If there is a regression then please provide a test that demonstrates it.

Comment (uklotzde) [reply src/track/serato/tags.cpp:24]: The type "ogg" will always map to "ogg" if this is one of the file suffixes provided by QMimeDatabase. I have even written a test that verifies this assumption!

We cannot test all possible situations for files with wrong suffixes. Use at you own risk or fix your file suffixes. I am done here.

Comment (daschuer) [reply src/track/serato/tags.cpp:24]: This PR scans for now for these files: 
"*.3g2", "*.3gp2", "*.3gpp2", "*.3gp", "*.3gpp", "*.3ga", "*.aac", "*.adts", "*.ass", "*.aiff", "*.aif", "*.flac", "*.it", "*.m4a", "*.f4a", "*.mp4", "*.m4v", "*.f4v", "*.lrv", "*.mod", "*.ult", "*.uni", "*.m15", "*.mtm", "*.669", "*.med", "*.mj2", "*.mjp2", "*.qt", "*.mov", "*.moov", "*.qtvr", "*.mp3", "*.mpga", "*.oga", "*.ogg", "*.opus", "*.ogv", "*.s3m", "*.stm", "*.wav", "*.wv", "*.wvp", "*.xm"

before it was 
"*.3g2", "*.3gp", "*.aac", "*.aif", "*.aiff", "*.caf", "*.flac", "*.it", "*.m4a", "*.m4v", "*.med", "*.mj2", "*.mod", "*.mov", "*.mp3", "*.mp4", "*.ogg", "*.okt", "*.opus", "*.s3m", "*.stm", "*.wav", "*.wv", "*.xm"

Now for instance a "*.3gpp2" file or an "*.oga" is listed in the library, but it is not playable. My guess is, that all new extensions are affected. 

If you do not have interest to solve this yourself, I can offer to issue a PR against your branch. But let's discuss fist which solution will be acceptable. I see the options already outlined above: 
* Probe all sibiling suffixes when selecting the soundsource 
* Maintain a lookup table with the correct FileType string per mimeType 
* Replace FileType by QMimeTime. 

What do you think?

**Comment by uklotzde** [comment]
Force pushing the test extensions until CI passes.

File: src/track/serato/tags.cpp
Comment (uklotzde) [reply src/track/serato/tags.cpp:24]: No. Either merge this PR or prove me wrong with a regression test that previously worked.

Comment (daschuer) [reply src/track/serato/tags.cpp:24]: I will not do the effort of writing a test. You can easily test it manually. Just rename a ogg file to oga. Rescan your library, the new fille will show up with FilyType oga. Now try to play it -> does not work. 

This is not a regression, but an incomplete feature of this branch, I consider as a bug. 

But anyway, I am happy to help. Did you consider the three options for a solution? Which one do you like to merge?

Comment (JoergAtGithub) [reply src/track/serato/tags.cpp:24]: If it's not a regression, it must not block merging this PR!
Because this PR fixes other issue and doesn't introduce any new bug. Any further improvements can be done in follow-up PRs.

Comment (uklotzde) [reply src/track/serato/tags.cpp:24]: @JoergAtGithub Thank you!

Hardly anyone will be affected by the remaining shortcomings and I refuse to spend more time on this.

Comment (daschuer) [reply src/track/serato/tags.cpp:24]: Which bug is fixed by this PR? 

From the user perspective it finds new files when scanning the library, but they cannot be played. The bad thing in this is that, these new files are persistent in the database. So we must not merge this unless we are willing to provide a clean up routine in a follow up PR.

But back to topic. I have already offered to bring this in a merge-able state. So let's look into the possibles solutions? Which one will fit?  I need that info because I have no interest to do one version and change it to another in a code review.

**Pull request closed without merging**
